Include/pystate.h

```diff
@@ -185,6 +204,74 @@ typedef struct _ts {
 } PyThreadState;
 #endif
 
+#ifndef Py_LIMITED_API
+
+typedef struct _frame *(*PyThreadFrameGetter)(PyThreadState *self_);
+
+struct _gilstate_globals {
+    int check_enabled;
+#ifdef Py_BUILD_CORE
+    /* Assuming the current thread holds the GIL, this is the
+       PyThreadState for the current thread. */
+    _Py_atomic_address tstate_current;
+#endif
+    PyThreadFrameGetter getframe;
+#ifdef WITH_THREAD
+    /* The single PyInterpreterState used by this process'
+       GILState implementation
+    */
+    /* TODO: Given interp_main, it may be possible to kill this ref */
+    PyInterpreterState *autoInterpreterState;
+    int autoTLSkey;
+#endif /* WITH_THREAD */
+};
+
+#include "warnings.h"
+#include "ceval.h"
+
+typedef struct pyruntimestate {
+    int initialized;
+    int core_initialized;
+    PyThreadState *finalizing;
+
+    struct pyinterpreters {
+#ifdef WITH_THREAD
+#ifdef Py_BUILD_CORE
```

Comment:
Having these `Py_BUILD_CORE` conditionals inside the struct definitions doesn't look right to me - don't they mean that the compiler will generate incorrect field offsets when compiled *without* `Py_BUILD_CORE` set?

Even having a `Py_BUILD_CORE` block at the *end* of the struct definitions seems problematic, since we later compose them all into a single struct, where the non-core builds would again get the field sizes wrong.

So if we're going to have a distinction between "public" runtime state (accessible via macros in the public C API) and "private" runtime state (accessible only when "Py_BUILD_CORE" is set), it seems to me that distinction is going to need to be handled by having two different top level runtime structs (one public and one private), and having a similar split for every section that includes non-public process global state.